Sources/SwiftGRPCNIO/ClientCalls/BaseClientCall.swift

```diff
@@ -0,0 +1,186 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Foundation
+import NIO
+import NIOHTTP1
+import NIOHTTP2
+import SwiftProtobuf
+
+/// This class provides much of the boilerplate for the four types of gRPC call objects returned to framework
+/// users.
+///
+/// Setup includes:
+/// - creation of an HTTP/2 stream for the call to execute on,
+/// - configuration of the NIO channel handlers for the stream, and
+/// - setting a call timeout, if one is provided.
+///
+/// This class also provides much of the framework user facing functionality via conformance to `ClientCall`.
+open class BaseClientCall<RequestMessage: Message, ResponseMessage: Message> {
+  /// The underlying `GRPCClient` providing the HTTP/2 channel and multiplexer.
+  internal let client: GRPCClient
+
+  /// Promise for an HTTP/2 stream to execute the call on.
+  internal let streamPromise: EventLoopPromise<Channel>
+
+  /// Client channel handler. Handles internal state for reading/writing messages to the channel.
+  /// The handler also owns the promises for the futures that this class surfaces to the user (such as
+  /// `initialMetadata` and `status`).
+  internal let clientChannelHandler: GRPCClientChannelHandler<RequestMessage, ResponseMessage>
+
+  /// Sets up a gRPC call.
+  ///
+  /// A number of actions are performed:
+  /// - a new HTTP/2 stream is created and configured using the channel and multiplexer provided by `client`,
+  /// - a callback is registered on the new stream (`subchannel`) to send the request head,
+  /// - a timeout is scheduled if one is set in the `callOptions`.
+  ///
+  /// - Parameters:
+  ///   - client: client containing the HTTP/2 channel and multiplexer to use for this call.
+  ///   - path: path for this RPC method.
+  ///   - callOptions: options to use when configuring this call.
+  ///   - responseObserver: observer for received messages.
+  init(
+    client: GRPCClient,
+    path: String,
+    callOptions: CallOptions,
+    responseObserver: ResponseObserver<ResponseMessage>
+  ) {
+    self.client = client
+    self.streamPromise = client.channel.eventLoop.newPromise()
+    self.clientChannelHandler = GRPCClientChannelHandler(
+      initialMetadataPromise: client.channel.eventLoop.newPromise(),
+      statusPromise: client.channel.eventLoop.newPromise(),
+      responseObserver: responseObserver)
+
+    self.createStreamChannel()
+    self.setTimeout(callOptions.timeout)
+
+    let requestHead = BaseClientCall<RequestMessage, ResponseMessage>.makeRequestHead(path: path, host: client.host, callOptions: callOptions)
+    self.sendHead(requestHead)
```

Comment:
In theory, it would be nice to also chain `sendMessage` and `sendEnd` off this `sendHead`. However, I guess that would make our interface more complicated (this would probably need to be a factory method returning an `EventLoopFuture<BaseClientCall>`). So I guess not chaining here should be fine. WDYT?

Reply:
Well done!